### PR TITLE
feat(cast): simulate published transaction locally

### PR DESCRIPTION
**Pull request by joshieDo**
## Motivation

Run a published transaction locally with trace and/or debug.

## Solution

* Fork (transaction.block - 1)
* Load and execute all transactions previous to ours in the transaction.block
* Execute ours
* Decode with EtherscanIdentifier

```
cast-run 
Runs a published transaction in a local environment.

USAGE:
    cast run [OPTIONS] --rpc-url <RPC_URL> <TX>

ARGS:
    <TX>    The transaction hash.

OPTIONS:
    -d, --debug                Debugs the transaction.
    -h, --help                 Print help information
        --label <LABEL>        Labels address in the trace.
                               0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045:vitalik.eth
    -q, --quick                Executes the transaction only with the state from the previous block. May
                               result in different results than the live execution!
    -r, --rpc-url <RPC_URL>    [env: ETH_RPC_URL=https://rpc.flashbots.net
```

Example (BEAN exploit):
```
cast run 0xcd314668aaa9bbfebaf1a0bd2b6553d01dd58899c508d4729fa7311dc5d33ad7 --rpc-url $ARCHIVE_RPC
```
image

too long, didn't fit lol

### CODE REVIEW

File: cli/src/cmd/forge/sim.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    #[clap(flatten, next_help_heading = "BUILD OPTIONS")]
+    pub opts: CoreBuildArgs,
+
+    #[clap(flatten, next_help_heading = "EVM OPTIONS")]
+    pub evm_opts: EvmArgs,
+
+    pub tx: String,
+
+    #[clap(long)]
+    pub rpc: String,
```
Comment (joshieDo) [inline cli/src/cmd/forge/sim.rs:46]: redundant, but how can i make the one from `evm_opts`  mandatory?

**Comment by mds1** [comment]
This is awesome, I was excited for this functionality!

Though I think this should be part of `cast` instead. https://github.com/foundry-rs/foundry/issues/29 is where `cast run-tx` (analogous to `seth run-tx`) is tracked, and that's pretty much exactly what this is

**Comment by joshieDo** [comment]
`forge run-tx` :sweat_smile: ?

I thought about including it on `cast` but in the end i decided for `forge` because of this:
```rust
use forge::{
    debug::DebugArena,
    decode::decode_console_logs,
    executor::{builder::Backend, opts::EvmOpts, DeployResult, ExecutorBuilder, RawCallResult},
    trace::{identifier::EtherscanIdentifier, CallTraceArena, CallTraceDecoderBuilder, TraceKind},
};
```

The dependencies are mostly in the `forge` crate, and making it accessible on `cast` is messy no?

**Comment by onbjerg** [comment]
Executor/trace/debug are re-exports of the evm crate

**Comment by mds1** [comment]
Should we call it `cast run-tx`, or just `cast run`? I kind of like the shorter version, and it parallels `forge run` nicely.

Also @joshieDo no pressure at all to implement all of the below in this PR, but figured this is a good spot to document the list of features I liked from `seth run-tx` + what I wish seth had. I can move these all into a separate issue depending on how much is done here. But it's about time I get them out of my head and documented somewhere (I thought I did this somewhere but can't find it) 😅
- `cast run <tx> --trace` to show a trace like you have above
- `cast run <tx> --debug` to launch the tx in the debugger (these first two are analogous to seth)
- `--quick` skips simulating previous txs, for speed (docs should note this is unsound and may give different results than the live tx). Also how slow was simulating previous txs for you? In seth it was painfully slow so I always skipped it
- If a contract name is verified on etherscan, use it's name in the trace (and append IDs if there's more than one with the same name, e.g. `Vyper_contract-1`, `Vyper_contract-2`). Seems this is partly done already 🎉
- `--label <addr1> <name1>  --label <addr2> <name2>` to give nice labels to arbitrary addresses
- `--dense` flag to hide addresses for labeled contracts (can only hide them entirely if we ensure unique labels, or just shrinking to `0x123...def` works too
- replace numbers with more human friendly representations, e.g. lots of numbers in that screenshot can be replaced with `2^255 - 1`

**Comment by gakonst** [comment]
+1 on "cast run" and using the foundry_evm exports

**Comment by joshieDo** [comment]
Moved it and simplified some stuff

Thanks @mds1 , i like those suggestions

> cast run <tx> --trace to show a trace like you have above

default behaviour has tracing, so I didn't include the flag. What does `seth` show if no flag is set?

> --quick skips simulating previous txs, for speed (docs should note this is unsound and may give different results than the live tx). Also how slow was simulating previous txs for you? In seth it was painfully slow so I always skipped it

Added the `--quick`. Regarding the slowness... it depends on how many transactions there are in the block and the RPC responsiveness , but I have a feeling the provider is caching the requests, so it should happen only once. Emulating the actual transactions on the EVM is pretty damn fast in my experience.

> --label <addr1> <name1> --label <addr2> <name2> to give nice labels to arbitrary addresses

I went for the format `address:label`

--

All the other suggestions, should probably done in other issues, since it touches on how traces are displayed, if I'm not mistaken.

**Comment by joshieDo** [comment]
Example of the speed:

Not cached:
```shell
time cargo run --bin cast -- run 0xa8faf11f4e4c0a93fe9df333cc18d15e854d94b9f19d2237899213cf73c4699a --rpc-url $RPC
    Finished dev [unoptimized + debuginfo] target(s) in 0.38s
     Running `target/debug/cast run 0xa8faf11f4e4c0a93fe9df333cc18d15e854d94b9f19d2237899213cf73c4699a --rpc-url $RPC
Executing previous transactions from the block.
Traces:
  [0] 0x165c…5e14::fallback{value: 750000000000000000000}() 
    └─ ← ()


Script ran successfully.
Gas used: 21000
cargo run --bin cast -- run  --rpc-url   2.61s user 0.25s system 4% cpu 1:01.39 total
```
Cached:
```shell
time cargo run --bin cast -- run 0xa8faf11f4e4c0a93fe9df333cc18d15e854d94b9f19d2237899213cf73c4699a --rpc-url $RPC
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
     Running `target/debug/cast run 0xa8faf11f4e4c0a93fe9df333cc18d15e854d94b9f19d2237899213cf73c4699a --rpc-url $RPC
Executing previous transactions from the block.
Traces:
  [0] 0x165c…5e14::fallback{value: 750000000000000000000}() 
    └─ ← ()


Script ran successfully.
Gas used: 21000
cargo run --bin cast -- run  --rpc-url   1.24s user 0.13s system 40% cpu 3.399 total
```

**Comment by joshieDo** [comment]
opening for review

I was thinking of maybe looking into debugging with the source code mapping in another PR, but open to include it in this one

**Comment by mds1** [comment]
Working great!! Thanks a lot @joshieDo 🔥

Only issue I can find is that if a tx fails in the constructor, you get a panic instead of a trace, but this is present in forge too, so not a problem caused by your PR. @onbjerg I recall we've briefly discussed this in an issue before, but not sure if there's a dedicated issue for this? If so let me know and I can create one 

```
cast run 0xcd314668aaa9bbfebaf1a0bd2b6553d01dd58899c508d4729fa7311dc5d33ad7 -q
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value:
   0: deployment failed: Revert

Location:
   ~/.foundry/foundry-rs/foundry/evm/src/executor/mod.rs:456

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
Location: cli/src/cmd/cast/run.rs:116

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
[1]    37788 abort      cast run 0xcd314668aaa9bbfebaf1a0bd2b6553d01dd58899c508d4729fa7311dc5d33ad7 -
```

File: cli/src/cmd/cast/run.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            let env = runtime.block_on(evm_opts.evm_env());
+            let db = runtime.block_on(Backend::new(
+                utils::get_fork(&evm_opts, &config.rpc_storage_caching),
+                &env,
+            ));
+
+            let builder = ExecutorBuilder::new()
+                .with_config(env)
+                .with_spec(crate::utils::evm_spec(&config.evm_version))
+                .with_gas_limit(evm_opts.gas_limit());
```
Comment (mds1) [inline cli/src/cmd/cast/run.rs:73]: Probably want to use the txs gas limit with `tx.gas` here instead?

File: cli/src/opts/cast.rs [hunk 3]
```diff
@@ -619,6 +619,8 @@ If an address is specified, then the ABI is fetched from Etherscan."#
         #[clap(arg_enum)]
         shell: clap_complete::Shell,
     },
+    #[clap(name = "run", about = "Runs a published transaction in a local environment.")]
```
Comment (mds1) [inline cli/src/opts/cast.rs:622]:
```suggestion
    #[clap(name = "run", about = "Runs a published transaction in a local environment and prints the trace.")]
```

**Review by mattsse — CHANGES_REQUESTED** [review]
some nits

File: cli/src/cmd/cast/run.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                println!("Executing previous transactions from the block.");
+
+                let block_txes = provider.get_block_with_txs(tx_block_number).await?;
+
+                for past_tx in block_txes.unwrap().transactions.into_iter() {
+                    if past_tx.hash().eq(&tx_hash) {
+                        break
+                    }
+
+                    executor.set_gas_limit(tx.gas);
```
Comment (mds1) [inline cli/src/cmd/cast/run.rs:91]: Looks like this is setting the gas limit of past txs to our txs gas limit

```suggestion
                    executor.set_gas_limit(past_tx.gas);
```

**Review by gakonst — APPROVED** [review]
LGTM. Let's ship it and if there's any bugs iterate from there?
@joshieDo this is a fun enough feature, feel free to tweet it out to the world :)

**Pull request merged**